py/gc.c

```diff
@@ -421,14 +429,18 @@ STATIC void gc_sweep(void) {
                         memset((void *)PTR_FROM_BLOCK(area, block), 0, BYTES_PER_BLOCK);
                         #endif
                     }
+                    last_block = block;
```

Comment:
Why is this line needed?  In this case here the sweep is collecting the head/tail block(s) and converting them to free blocks, so why should the block be recorded as the last block?

Reply:
I think this line is necessary but should be inside an `else`. So if we're freeing the tail, then don't record it as the last block, but if we're not then this is an (effectively unmarked) tail, so it does need to be recorded.

Also, I think we should rename `last_block` to `last_in_use_block`.